lightning/src/events/bump_transaction.rs

```diff
@@ -245,3 +259,383 @@ pub enum BumpTransactionEvent {
+/// [`CoinSelectionSource`] to fee bump transactions via Child-Pays-For-Parent (CPFP) or
+/// Replace-By-Fee (RBF).
+pub struct BumpTransactionEventHandler<B: Deref, C: Deref, SP: Deref, L: Deref>
+where
+	B::Target: BroadcasterInterface,
+	C::Target: CoinSelectionSource,
+	SP::Target: SignerProvider,
+	L::Target: Logger,
+{
+	broadcaster: B,
+	utxo_source: C,
+	signer_provider: SP,
+	logger: L,
+	secp: Secp256k1<secp256k1::All>,
+}
+
+impl<B: Deref, C: Deref, SP: Deref, L: Deref> BumpTransactionEventHandler<B, C, SP, L>
+where
+	B::Target: BroadcasterInterface,
+	C::Target: CoinSelectionSource,
+	SP::Target: SignerProvider,
+	L::Target: Logger,
+{
+	/// Returns a new instance capable of handling [`Event::BumpTransaction`] events.
+	pub fn new(broadcaster: B, utxo_source: C, signer_provider: SP, logger: L) -> Self {
+		Self {
+			broadcaster,
+			utxo_source,
+			signer_provider,
+			logger,
+			secp: Secp256k1::new(),
+		}
+	}
+
+	/// Updates a transaction with the result of a successful coin selection attempt.
+	fn process_coin_selection(
+		&self, tx: &mut Transaction, mut coin_selection: CoinSelection,
+		mut override_change_output: Option<impl FnOnce(&mut Transaction, &mut CoinSelection)>,
+	) {
+		for utxo in coin_selection.confirmed_utxos.drain(..) {
+			tx.input.push(TxIn {
+				previous_output: utxo.outpoint,
+				script_sig: Script::new(),
+				sequence: Sequence::ZERO,
+				witness: Witness::new(),
+			});
+		}
+		if let Some(override_change_output) = override_change_output.take() {
+			override_change_output(tx, &mut coin_selection)
+		} else if let Some(change_output) = coin_selection.change_output.take() {
+			tx.output.push(change_output);
+		}
+	}
+
+	/// Returns an unsigned transaction spending an anchor output of the commitment transaction, and
+	/// any additional UTXOs sourced, to bump the commitment transaction's fee.
+	fn build_anchor_tx(
+		&self, claim_id: ClaimId, target_feerate_sat_per_1000_weight: u32,
+		commitment_tx: &Transaction, anchor_descriptor: &AnchorDescriptor,
+	) -> Result<Transaction, ()> {
+		let must_spend = vec![Input {
+			outpoint: anchor_descriptor.outpoint,
+			witness_weight: commitment_tx.weight() as u64 + ANCHOR_INPUT_WITNESS_WEIGHT,
+		}];
+		let coin_selection = self.utxo_source.select_confirmed_utxos(
+			claim_id, &must_spend, &[], target_feerate_sat_per_1000_weight,
+		)?;
+		let override_change_output = |tx: &mut Transaction, coin_selection: &mut CoinSelection| {
+			if let Some(change_output) = coin_selection.change_output.take() {
+				tx.output.push(change_output);
+			} else {
+				// We weren't provided a change output, likely because the input set was a perfect
+				// match, but we still need to have at least one output in the transaction for it to
+				// be considered standard. We choose to go with an empty OP_RETURN as it is the
+				// cheapest way to include a dummy output.
+				tx.output.push(TxOut {
+					value: 0,
+					script_pubkey: Script::new_op_return(&[]),
```

Comment:
So we’re exposed to our anchor CPFP not propagating if `datacarrier` is turn off by the node operator itself, or become a widely deployed policy on the transaction-relay network in the future on the ground it’s only used to carry on zero-value transaction ?

What are doing other implementations in this case ? If there is a common behavior, that might be rational for fee saving purposes, at least it would be better for this to be documented.

Reply:
This is a consensus rule to have at least one output not standard rule, see `CheckTransaction()`.

Reply:
Still would be nice if we don’t forget in the anchor documentation somewhere to say “YOU SHOULD NOT TURN OFF `datacarrier` OPTION ON YOUR FULL_NODE”.

> They don't have this problem since they're tightly integrated with their on-chain wallet. They just add a regular change output.

This approach is more fee-saving, as op_return are provably unspendale and therefore don’t have to satisfy dust outputs checks, which are even more a mess.

Note, as the CPFP should always at least 2 inputs, I think we’re always good with `MIN_STANDARD_TX_NONWITNESS_SIZE`.